### PR TITLE
Overlay index fetch issue

### LINKED ISSUE #78: Overlay with Fetch/IndexedDB backend issue
Hi, 

I'm using zenfs to implement an Overlay Filesystem with a Fetch/IndexedDB readable/writable backend. I am having a bit of trouble using the filesystem/understanding the documentation.

I'm first building the index.json file by using `npx make-index`. From there, I configure the filesystem and attempt to read a file using fs.readFile. I'm receiving an error that the file does not exist despite the fact that I prepopulated the filesystem with the file. Do you have any tips on how I can resolve this?

I have a minimum working example at this github repo: https://github.com/terryluan12/zen-fs-example.git

### PR DESCRIPTION
**Pull request by terryluan12**
Should resolve #78 and one of the additional bugs mentioned in #80

### CODE REVIEW

**Comment by terryluan12** [comment]
As mentioned in #80, there were issues when initializing the `Stat` objects. Looking further into it, I found that the `dirEntries` was called to get the directory entries. However, it simply iterates through the keys of the `Index`. Since the `Index` object is in the process of being created when this is called, not all keys which would otherwise be in the final `Index` object are there when it's called.

I also found the bug which caused the double unlocking bug. This occurs if the program locks a path 3 times. 
On it's first lock, it would creates a Promise 'A'. Then when it locks a second time, it `awaits` Promise 'A'. Finally on it's third time, it would also `await` Promise 'A'. When the program unlocks the lock (resolving Promise 'A'), both threads which are awaiting 'A' start running at the same time, which causes the bug.

**Comment by james-pre** [comment]
@terryluan12 Some changes have been made. Primarily, the `Mutex` logic has been folded into `LockedFS`. Please consult the release notes for v0.14.0.

**Comment by terryluan12** [comment]
Ah, didn't even notice it, thanks for the heads up! I just merged + re-added the changes

File: src/backends/index/index.ts
Comment (james-pre) [inline src/backends/index/index.ts:1]: These changes relate to the issue you mentioned in #80, please refer to my other comment.

File: src/backends/locked.ts [hunk 1]
```diff
@@ -26,7 +26,7 @@ export class LockedFS<FS extends FileSystem> implements FileSystem {
 	/**
 	 * The current locks
 	 */
-	private locks: Map<string, MutexLock> = new Map();
+	private locks: Map<string, [MutexLock]> = new Map();
```
Comment (james-pre) [inline src/backends/locked.ts:29]: This is a tuple with a single `MutexLock`. Do you mean an array, `MutexLock[]`?

File: src/backends/locked.ts [hunk 2]
```diff
@@ -47,7 +51,8 @@ export class LockedFS<FS extends FileSystem> implements FileSystem {
 	public async lock(path: string): Promise<MutexLock> {
 		if (this.locks.has(path)) {
 			// Non-null assertion: we already checked locks has path
-			await this.locks.get(path)!.promise;
```
Comment (james-pre) [inline src/backends/locked.ts:50]: 1. This is not a promise. `await`ing the lock has no effect and is a bug.
2. Please inline: `await this.locks.get(path)!.at(-1).promise`

File: src/backends/locked.ts [hunk 3]
```diff
@@ -83,8 +88,8 @@ export class LockedFS<FS extends FileSystem> implements FileSystem {
 		}
 
 		// Non-null assertion: we already checked locks has path
-		this.locks.get(path)!.resolve();
-		this.locks.delete(path);
+		const res = this.locks.get(path)!.shift();
```
Comment (james-pre) [inline src/backends/locked.ts:91]: Again, please inline: `this.locks.get(path)!.shift().resolve();`

File: src/backends/locked.ts [hunk 4]
```diff
@@ -93,7 +98,7 @@ export class LockedFS<FS extends FileSystem> implements FileSystem {
 	 * @internal
 	 */
 	public isLocked(path: string): boolean {
-		return this.locks.has(path);
+		return this.locks.has(path) && this.locks.get(path)!.length > 0;
```
Comment (james-pre) [inline src/backends/locked.ts:101]: Can be simplified: `return !!this.locks.get(path)?.length`

**Review by james-pre — CHANGES_REQUESTED** [review]

**Comment by james-pre** [comment]
> As mentioned in #80, there were issues when initializing the `Stat` objects. Looking further into it, I found that the `dirEntries` was called to get the directory entries. However, it simply iterates through the keys of the `Index`. Since the `Index` object is in the process of being created when this is called, not all keys which would otherwise be in the final `Index` object are there when it's called.

This should have been resolved in https://github.com/zen-fs/core/commit/ec86a6aacfcc36d0960f2a1280b88d21bce8bc92. Please let me know if the issue still happens- you will need to run `make-index` again. I am again using a regex since it is relatively slow. If one was to have 10,000 files for example, creating a new regex for each one slows things down.

**Comment by james-pre** [comment]
@terryluan12 Any updates on this?

**Comment by james-pre** [comment]
@terryluan12 Any updates on this?

**Comment by terryluan12** [comment]
Hi @james-pre, so sorrry about the lack of update! Got busy with some real life stuff, and didn't see the comments. Just got a replacement for my previous laptop, and will get to work on zen-fs, and finish up this issue ASAP

**Comment by terryluan12** [comment]
Hey @james-pre, I've been taking a look at the double unlocking issue, and I've gotten an idea of why it's bugging. Specifically, it occurs when you have two processes waiting on a locked promise. Once the promise resolves, both processes will continue at the same time. Thus they'll both call addLock, and one promise will override the others, causing the weird double unlocking. Thinking about it, I think one possible solution is to implement a queuing system. What are your thoughts?

**Comment by james-pre** [comment]
@terryluan12,

Thanks for your dedication to this.

My thinking is that we should try to avoid a queue system.

It *should* work like this:
- call asynchronous methods `a` and `b` on the same path without using `await`,
	- i.e.
	```ts
		a(path);
		b(path);
	```
- `a` `lock`s on the path, which adds the lock
- `b` tries to `lock` on the path, though it must wait for `a` to finish (in `LockedFS.lock`)
- `a` performs its operations
- `a`'s scope is left, `MutexLock[Symbol.dispose]` runs, unlocking the path
- `b` then performs its operations
-  `b`'s scope is left, `MutexLock[Symbol.dispose]` runs, unlocking the path

I think https://github.com/zen-fs/core/commit/e3878895f107ed43e412b226f4feab6232123a54 may fix something since now the promise won't resolve until the lock is actually cleared.

**Comment by terryluan12** [comment]
No worries at all, thanks for allowing me to help! 

I think there's been a misunderstanding. To clarify, the issue occurs in the following scenario, when there are 3 locks; for example A, B, and C. 
In this case, if A locks first, and then B and C lock (in any order), then both B and C are now waiting on A's lock to release. 

When A then releases, both B and C start again at the same time. I believe this is the unintended behaviour, where they both start again at the same time.

**Comment by james-pre** [comment]
Ah, this makes more sense now! Thanks for the clarification.

Maybe we could look at a wrap and replace, so when B attempts the lock, it replaces the current lock with one that resolves when B is done. This wouldn't impact A since it holds the lock, and doesn't update the reference (i.e. the A's `MutexLock[Symbol.dipose]` resolves only the lock for A).

**Comment by james-pre** [comment]
@terryluan12 Now that the mutex issue has been solved, can I close this PR?

**Comment by terryluan12** [comment]
Yep, absolutely!

**Pull request closed without merging**